### ISSUE TITLE
Cannot map properties defined in interfaces

### ISSUE THREAD

**Issue by ChristophGr** [modelmapper/modelmapper#144]
When the getter for a Property is defined in an Interface, it cannot be mapped because of 

```
Caused by: java.lang.NullPointerException
    at org.modelmapper.internal.ExplicitMappingVisitor$MappingCapturingVisitor.assertNotFinal(ExplicitMappingVisitor.java:322)
    at org.modelmapper.internal.ExplicitMappingVisitor$MappingCapturingVisitor.recordSourceMethod(ExplicitMappingVisitor.java:263)
    at org.modelmapper.internal.ExplicitMappingVisitor$MappingCapturingVisitor.visitEnd(ExplicitMappingVisitor.java:207)
    at org.modelmapper.internal.asm.ClassReader.b(Unknown Source)
    at org.modelmapper.internal.asm.ClassReader.accept(Unknown Source)
    at org.modelmapper.internal.asm.ClassReader.accept(Unknown Source)
    at org.modelmapper.internal.ExplicitMappingBuilder.visitPropertyMap(ExplicitMappingBuilder.java:222)
    at org.modelmapper.PropertyMap.configure(PropertyMap.java:380)
    ... 44 more
```

**Comment by jhalterman** [comment]
Thanks for this. Do you have an example of where this pops up? I tried reproducing, but I'm probably not getting your use case correct:

```
@Test
public class GH145 extends AbstractTest {
  interface ISrc {
    String getName();
  }
  
  static class Src implements ISrc {
    String name;
    public String getName() {
      return name;
    }
  }
  
  interface IDest {
    void setName(String name);
  }
  
  static class Dest implements IDest {
    String name;
    public void setName(String name) {
      this.name = name;
    }
  }

  @Test
  public void shouldMapCircularReferencesWithProvider() {
    Src source = new Src(); 
    source.name = "foo";
    Dest dest = modelMapper.map(source, Dest.class);
    
    assertEquals(dest.name, source.name);
  }
}

```

**Comment by ChristophGr** [comment]
```
interface Parent {
    String getName();
}

interface DerivedInterface extends Parent {
}

static class Source {
    DerivedInterface sourceField;
}
```

The problem is, that calling getSuperClass() on an extended interface returns null.

**Comment by DouglasBrunner** [comment]
I believe I'm facing a similar issue to the original poster.

When I pass an interface to the PropertyMap ModelMapper throws a null pointer exception, however, when I pass in the class which implements the interface I get the expected result.

Maybe I'm misunderstanding the intended usage of the PropertyMap, but it seems like we should be able to program to interfaces so raised the issue here.

Here's the code.

**Main.java**
```java
import org.modelmapper.ModelMapper;

public final class Main {
    public static void main(String[] args) {

        ModelMapper mm = new ModelMapper();
        mm.addMappings(new Mappings());
        DestinationModel dest = new DestinationModel();
        mm.map(new SourceModel(), dest);
        System.out.println("Name: " + dest.getName());
        System.out.println("AGE: " + dest.getNestedModel().getAge());
    }
}
```

**DestinationModel.java**
```java
public class DestinationModel {
    private String name;
    private NestedDestinationModel nestedModel;

    public String getName() {
        return name;
    }

    public void setName(final String name) {
        this.name = name;
    }

    public NestedDestinationModel getNestedModel() {
        return nestedModel;
    }

    public void setNestedModel(final NestedDestinationModel nestedModel) {
        this.nestedModel = nestedModel;
    }

}
```

**NestedDestinationModel.java**
```java
public class NestedDestinationModel {
    private String age;

    public String getAge() {
        return age;
    }

    public void setAge(final String age) {
        this.age = age;
    }
}
```

**SourceModel.java**
```java
public class SourceModel implements MappingInterface {
    public String getName() {
        return "Bob Alice";
    }

    public String getAge() {
        return "99";
    }
}
```

**MappingInterface.java**
```java
public interface MappingInterface {
    public String getName();
    public String getAge();
}
```

**Mappings.java** - *Working Version*
```java
import org.modelmapper.PropertyMap;

public class Mappings extends PropertyMap<SourceModel, DestinationModel> {
    protected void configure() {
        map().setName(source.getName());
        map().getNestedModel().setAge(source.getAge());
    }
}
```

**Mappings.java** - *NullPointer Version*
```java
import org.modelmapper.PropertyMap;

public class Mappings extends PropertyMap<MappingInterface, DestinationModel> {
    protected void configure() {
        map().setName(source.getName());
        map().getNestedModel().setAge(source.getAge());
    }
}
```

**Comment by chhsiao90** [comment]
@DouglasBrunner -

I think your issue is similar but different to this issue.
The original issue seems had been fixed by #145 .
I'm not sure why this issue keep open!

So I think it would be better to create a new issue and reference to this issue.
To prevent some kind of misleading...

BTW, after investigation, I can point out that this issue related to [MappingEngineImpl#L89](https://github.com/jhalterman/modelmapper/blob/master/core/src/main/java/org/modelmapper/internal/MappingEngineImpl.java#L89)
```java
    TypeMap<S, D> typeMap = typeMapStore.get(context.getSourceType(), context.getDestinationType(),
        context.getTypeMapName());
```

That for successful case, it can get correct TypeMap<SourceModel, DestinationModel> from typeMapStore
But failure case, it cannot get TypeMap<MappingInterface, DestinationModel> because it had different source type.

Currently, I have no idea how to fix it..

**Comment by chhsiao90** [comment]
I think this might be fixed by #145 so I closed it.
If still have problem, feel free to reopen!
